core/src/test/java/org/apache/iceberg/io/TestResolvingIO.java

```diff
@@ -47,4 +55,23 @@ public void testResolvingFileIOJavaSerialization() throws IOException, ClassNotF
     FileIO roundTripSerializedFileIO = TestHelpers.roundTripSerialize(testResolvingFileIO);
     assertThat(roundTripSerializedFileIO.properties()).isEqualTo(testResolvingFileIO.properties());
   }
+
+  @Test
+  public void testEnsureInterfaceImplementation() {
+    ResolvingFileIO testResolvingFileIO = spy(new ResolvingFileIO());
```

Comment:
I think this is ok, but what I was hoping we would have coverage for was that all delegates properly worked. This may be very difficult, I'm just looking for a way to confirm that the class will work at runtime. If we can force all the checks at compile time I'd be ok with that as well.

Reply:
We wouldn't be able to test the actual delegates without having the dependencies here (i.e. iceberg-aws and iceberg-gcp). I'll see if a service pattern could work where the delegate types register themselves with ResolvingFileIO, then we can have the compile time check.

Reply:
I looked into using Java SPI, but I don't think it is a good idea as it could cause problems with shadow jars. Also it isn't much different than this solution. We can't register with ResolvingFileIO from the delegate class as the delegate class might not be loaded at all yet.

One option, somewhat similar to https://github.com/apache/iceberg/pull/7976 would be to remove `SupportsPrefixOperations` and then provide a default implementation for bulk delete for the orphan file case so the delegate doesn't need the mixin interface. Though I was hoping we could have the prefix operations also as they won't be accessible at all.

Reply:
I added tests for each of the delegate types to ensure they load. Though that doesn't prevent a new class from being added to the list that doesn't implement the interface.